Fixed RFC7592 Dynamic Client Registration Management Protocol

azmeuk opened the pull request:
This PR fixes/implements the RFC7592 spec. It brings those changes

- The Flask integration keep the request data when the request type is `PUT`
- rfc7592/endpoint.py implements the RFC7592 specification
- The endpoint is tested by tests/flask/test_oauth2/test_client_configuration_endpoint.py
- rfc7592/endpoint.py has a 100% coverage
- The implementation is documented by docs/specs/rfc7592.rst

I read RFC7592 and did my best to implement this, but I am not 100% confident to have understood or implemented things right. I think a second pair of eyes might help. I tried to make unit tests as explicit and short as I could, and insert quotes from the spec to illustrate the tests.

The `extract_client_metadata` method has been copied from RFC7591 (and I removed the `software_statement` part). Maybe this should be factorized? If so, how should this be? I was thinking of extracting the method in a `ClientRegistrationEndpointMixin` from which would inherit `ClientRegistrationEndpoint` and `ClientConfigurationEndpoint`.

There are breaking changes on some method signatures, but this spec was not documented anyways:

- `introspect_client` MUST be implemented
- `save_client` is renamed `update_client` and takes an additional `client_metadata` parameter
- `introspect_metadata` is not required anymore
- `revoke_access_token` takes a `token` parameter

I think the documentation could make clearer that a *registration access token* can be a different thing than a regular *access_token*, but I guess this is enough like this.

Fixes #499

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

lepture commented on authlib/oauth2/rfc7592/endpoint.py:
```diff
@@ -145,28 +189,66 @@ def authenticate_token(self, request):
+        Developers MUST implement this method in subclass::
+
+            def authenticate_client(self, request):
+                return Client.query.get(request.data.get('client_id'))
```
please use a framework-free style example:

```py
            def authenticate_client(self, request):
                return query_client(request.data.get('client_id'))
```

lepture commented on authlib/oauth2/rfc7592/endpoint.py:
```diff
@@ -145,28 +189,66 @@ def authenticate_token(self, request):
         implement it in subclass, e.g.::
 
             def delete_client(self, client, request):
-                client.delete()
```
please use framework-free style example, `client.delete()` is better.

lepture commented:
1. I tried to remove sqlachemy from authlib.
2. are you using some code formatting tool? most of the code are using `'` instead of `"`, please don't change them.

lepture commented:
Thanks for your contribution, I'll check the specification implementation later.

azmeuk commented:
I have added commits for simple quotes and framework independent examples. I added commits to the PR so you can review the differences, but if you prefer (and for future fixes) I can squash the commit and force push.

What do you think about using a pre-commit configuration to check and enforce the code style for authlib? We could take inspiration from the [wtforms pre-commit configuration](https://github.com/wtforms/wtforms/blob/master/.pre-commit-config.yaml). I can open a PR if you wish.

lepture commented:
@azmeuk I don't like the black code style. Let's keep it the current way.

For testing, I'm using `self.assertEqual` and other `self.assertXXX` from unittest module. These method would print better error stack than the raw `assert`.

lepture commented on docs/changelog.rst:
```diff
@@ -15,6 +15,7 @@ Version 1.2.0
 - Use ``flask.g`` instead of ``_app_ctx_stack``, via :gh:`issue#482`.
 - Add ``headers`` parameter back to ``ClientSecretJWT``, via :gh:`issue#457`.
 - Always passing ``realm`` parameter in OAuth 1 clients, via :gh:`issue#339`.
+- Implemented RFC7592 Dynamic Client Registration Management Protocol, via :gh:`issue#499`.
```
```
via :gh:`PR#505`.
```

lepture commented on tests/flask/test_oauth2/test_client_configuration_endpoint.py:
```diff
@@ -0,0 +1,505 @@
+        return Client.query.filter_by(client_id=client_id).first()
+
+    def revoke_access_token(self, request, token):
+        token.revoked = True
```
Do we need to add:

```
db.session.add(token)
db.session.commit()
```

azmeuk replied:
Tests were passing without this, but I suppose this is needed anyways.

lepture commented on authlib/oauth2/rfc7592/endpoint.py:
```diff
@@ -70,66 +71,109 @@ def create_delete_client_response(self, client, request):
+    def extract_client_metadata(self, request):
+        json_data = request.data.copy()
+        options = self.get_claims_options()
+        claims = self.claims_class(json_data, {}, options, self.get_server_metadata())
```
You need to define:

```
def get_server_metadata(self):
    raise NotImplementedError()
```

lepture commented:
We are almost done. But there should be a get_server_metadata method implementation in the rfc documentation.

Thanks for your hard work.

azmeuk commented:
This is done, does this implementation example enough?

lepture commented:
Thanks.
